### PR TITLE
Fixes calculating resolved project reference to redirect for module resolution

### LINKED ISSUE #38711: Project reference with baseUrl (or paths) fails to build
**TypeScript Version:**  4.0.0-dev.20200521

**Search Terms:** TS2307 project reference baseUrl

**Code**
https://github.com/govizlora/project_reference_baseUrl

In brief, I have a monorepo where `pkg1` relies on `pkg2`.

In `pkg2`, I use `"baseUrl": "."` and the package exports a type:
```ts
export type { TheNum } from 'const';
```
In `pkg1` I import that type:
```ts
import type { TheNum } from 'pkg2'
```

**Expected behavior:**
When I run `tsc -b` under `pkg1`, no error is emitted.

**Actual behavior:**
When I run `tsc -b` under `pkg1`, I see:
```
../pkg2/build/index.d.ts:1:29 - error TS2307: Cannot find module 'const' or its corresponding type declarations.

1 export type { TheNum } from 'const';
```

**Playground Link:**  N/A since it involves multiple projects. There is a github repo that reproduces the issue: https://github.com/govizlora/project_reference_baseUrl

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/29918 while my issue is not solved by `tsc -b`

### PR DESCRIPTION
**Pull request by sheetalkamat**
When the containing file for module resolution is output from referenced project, but wasn't mapped from source of referenced project we weren't looking for the project reference options to resolve module names or type reference directives correctly.
Eg when module resolution found .d.ts file instead of .ts file from referenced project

So now if file is d.ts, apart from getting referenced project to redirect(which is checks if file is source file from reference project), we need to try and see if file is source from project reference.

Fixes #38711

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -989,22 +989,60 @@ namespace ts {
 
         return program;
 
-        function resolveModuleNamesWorker(moduleNames: string[], containingFile: string, reusedNames?: string[], redirectedReference?: ResolvedProjectReference) {
+        function resolveModuleNamesWorker(moduleNames: string[], containingFile: SourceFile, reusedNames: string[] | undefined) {
+            if (!moduleNames.length) return [];
```
Comment (sheetalkamat) [inline src/compiler/program.ts:993]: This ensures that we aren't calculating redirect reference for files that contain no modules.

File: src/compiler/program.ts [hunk 2]
```diff
@@ -989,22 +989,60 @@ namespace ts {
 
         return program;
 
-        function resolveModuleNamesWorker(moduleNames: string[], containingFile: string, reusedNames?: string[], redirectedReference?: ResolvedProjectReference) {
+        function resolveModuleNamesWorker(moduleNames: string[], containingFile: SourceFile, reusedNames: string[] | undefined): readonly ResolvedModuleFull[] {
+            if (!moduleNames.length) return emptyArray;
```
Comment (sheetalkamat) [inline src/compiler/program.ts:993]: This ensures we try to locate source of project reference file on .d.ts file only if it contains modules

File: src/compiler/program.ts [hunk 3]
```diff
@@ -989,22 +989,60 @@ namespace ts {
...
+            if (resultFromDts) return resultFromDts;
+
+            // If preserveSymlinks is true, module resolution wont jump the symlink
+            // but the resolved real path may be the .d.ts from project reference
+            // Note:: Currently we try the real path only if the
+            // file is from node_modules to avoid having to run real path on all file paths
+            if (!host.realpath || !options.preserveSymlinks || !stringContains(file.originalFileName, nodeModulesPathPart)) return undefined;
+            const realDeclarationFileName = host.realpath(file.originalFileName);
+            const readlDeclarationPath = toPath(realDeclarationFileName);
+            return readlDeclarationPath === file.path ? undefined : getRedirectReferenceForResolutionFromSourceOfProject(realDeclarationFileName, readlDeclarationPath);
```
Comment (andrewbranch) [inline src/compiler/program.ts:1030]:
```suggestion
            const readlDeclarationPath = toPath(realDeclarationFileName);
            return realDeclarationPath === file.path ? undefined : getRedirectReferenceForResolutionFromSourceOfProject(realDeclarationFileName, realDeclarationPath);
```

**Review by andrewbranch — APPROVED** [review]

File: src/compiler/program.ts (hunk 3 above)
Comment (sheetalkamat) [inline src/compiler/program.ts:1030]:
```suggestion
            const realDeclarationPath = toPath(realDeclarationFileName);
            return realDeclarationPath === file.path ? undefined : getRedirectReferenceForResolutionFromSourceOfProject(realDeclarationFileName, realDeclarationPath);
```

**Pull request merged**